File: problems/043/043.f90
PROGRAM F043

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    CHARACTER (LEN=10) :: MAIN_NUM_STR = "0123456789"
    CHARACTER (LEN=1)  :: TEMP_CHAR

    INTEGER (KIND=8), DIMENSION(100) :: PANDIGITAL_NUM_LIST = 0;
    INTEGER (KIND=8)   :: SUM_VAL = 0
    INTEGER (KIND=8)   :: MAIN_NUM

    INTEGER :: COUNT_VAL = 1
    INTEGER :: I, J, K, L, M, N, O, P, Q, R, TEMP_VAL

    LOGICAL :: DUPLICATE_FOUND

    DO I = 1, 10

        TEMP_CHAR           = MAIN_NUM_STR(10:10)
        MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
        MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
        MAIN_NUM_STR(8:8)   = MAIN_NUM_STR(7:7)
        MAIN_NUM_STR(7:7)   = MAIN_NUM_STR(6:6)
        MAIN_NUM_STR(6:6)   = MAIN_NUM_STR(5:5)
        MAIN_NUM_STR(5:5)   = MAIN_NUM_STR(4:4)
        MAIN_NUM_STR(4:4)   = MAIN_NUM_STR(3:3)
        MAIN_NUM_STR(3:3)   = MAIN_NUM_STR(2:2)
        MAIN_NUM_STR(2:2)   = MAIN_NUM_STR(1:1)
        MAIN_NUM_STR(1:1)   = TEMP_CHAR

        DO J = 1, 9

            TEMP_CHAR           = MAIN_NUM_STR(10:10)
            MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
            MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
            MAIN_NUM_STR(8:8)   = MAIN_NUM_STR(7:7)
            MAIN_NUM_STR(7:7)   = MAIN_NUM_STR(6:6)
            MAIN_NUM_STR(6:6)   = MAIN_NUM_STR(5:5)
            MAIN_NUM_STR(5:5)   = MAIN_NUM_STR(4:4)
            MAIN_NUM_STR(4:4)   = MAIN_NUM_STR(3:3)
            MAIN_NUM_STR(3:3)   = MAIN_NUM_STR(2:2)
            MAIN_NUM_STR(2:2)   = TEMP_CHAR

            DO K = 1, 8

                TEMP_CHAR           = MAIN_NUM_STR(10:10)
                MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
                MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
                MAIN_NUM_STR(8:8)   = MAIN_NUM_STR(7:7)
                MAIN_NUM_STR(7:7)   = MAIN_NUM_STR(6:6)
                MAIN_NUM_STR(6:6)   = MAIN_NUM_STR(5:5)
                MAIN_NUM_STR(5:5)   = MAIN_NUM_STR(4:4)
                MAIN_NUM_STR(4:4)   = MAIN_NUM_STR(3:3)
                MAIN_NUM_STR(3:3)   = TEMP_CHAR

                DO L = 1, 7

                    TEMP_CHAR           = MAIN_NUM_STR(10:10)
                    MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
                    MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
                    MAIN_NUM_STR(8:8)   = MAIN_NUM_STR(7:7)
                    MAIN_NUM_STR(7:7)   = MAIN_NUM_STR(6:6)
                    MAIN_NUM_STR(6:6)   = MAIN_NUM_STR(5:5)
                    MAIN_NUM_STR(5:5)   = MAIN_NUM_STR(4:4)
                    MAIN_NUM_STR(4:4)   = TEMP_CHAR

                    DO M = 1, 6

                        TEMP_CHAR           = MAIN_NUM_STR(10:10)
                        MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
                        MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
                        MAIN_NUM_STR(8:8)   = MAIN_NUM_STR(7:7)
                        MAIN_NUM_STR(7:7)   = MAIN_NUM_STR(6:6)
                        MAIN_NUM_STR(6:6)   = MAIN_NUM_STR(5:5)
                        MAIN_NUM_STR(5:5)   = TEMP_CHAR

                        DO N = 1, 5

                            TEMP_CHAR           = MAIN_NUM_STR(10:10)
                            MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
                            MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
                            MAIN_NUM_STR(8:8)   = MAIN_NUM_STR(7:7)
                            MAIN_NUM_STR(7:7)   = MAIN_NUM_STR(6:6)
                            MAIN_NUM_STR(6:6)   = TEMP_CHAR

                            DO O = 1, 4

                                TEMP_CHAR           = MAIN_NUM_STR(10:10)
                                MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
                                MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
                                MAIN_NUM_STR(8:8)   = MAIN_NUM_STR(7:7)
                                MAIN_NUM_STR(7:7)   = TEMP_CHAR

                                DO P = 1, 3

                                    TEMP_CHAR           = MAIN_NUM_STR(10:10)
                                    MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
                                    MAIN_NUM_STR(9:9)   = MAIN_NUM_STR(8:8)
                                    MAIN_NUM_STR(8:8)   = TEMP_CHAR

                                    DO Q = 1, 2

                                        TEMP_CHAR           = MAIN_NUM_STR(10:10)
                                        MAIN_NUM_STR(10:10) = MAIN_NUM_STR(9:9)
                                        MAIN_NUM_STR(9:9)   = TEMP_CHAR

                                        IF ((MAIN_NUM_STR(6:6) /= '5')         &
                                      .AND. (MAIN_NUM_STR(6:6) /= '0')) CYCLE

                                        IF (MOD(IACHAR(MAIN_NUM_STR(4:4))      &
                                        -       IACHAR('0'), 2) /= 0) CYCLE

                                        IF (MOD(IACHAR(MAIN_NUM_STR(3:3))      &
                                        +       IACHAR(MAIN_NUM_STR(4:4))      &
                                        +       IACHAR(MAIN_NUM_STR(5:5))      &
                                        -      (IACHAR('0') * 3), 3) /= 0) CYCLE

                                        READ (MAIN_NUM_STR(5:7), *) TEMP_VAL
                                        IF (MOD(TEMP_VAL, 7) /= 0) CYCLE

                                        READ (MAIN_NUM_STR(6:8), *) TEMP_VAL
                                        IF (MOD(TEMP_VAL, 11) /= 0) CYCLE

                                        READ (MAIN_NUM_STR(7:9), *) TEMP_VAL
                                        IF (MOD(TEMP_VAL, 13) /= 0) CYCLE

                                        READ (MAIN_NUM_STR(8:10), *) TEMP_VAL
                                        IF (MOD(TEMP_VAL, 17) /= 0) CYCLE

                                        READ (MAIN_NUM_STR, *) MAIN_NUM

                                        DUPLICATE_FOUND = .FALSE.

                                        DO R = 1, COUNT_VAL
                                            IF (PANDIGITAL_NUM_LIST(R)         &
                                            ==  MAIN_NUM) THEN

                                                DUPLICATE_FOUND = .TRUE.
                                                EXIT

                                            END IF
                                        END DO

                                        IF (.NOT. DUPLICATE_FOUND) THEN

                                            PANDIGITAL_NUM_LIST(COUNT_VAL)     &
                                            = MAIN_NUM

                                            SUM_VAL   = SUM_VAL   + MAIN_NUM
                                            COUNT_VAL = COUNT_VAL + 1

                                        END IF

                                    END DO
                                END DO
                            END DO
                        END DO
                    END DO
                END DO
            END DO
        END DO
    END DO

    PRINT ('(I0)'), SUM_VAL

END PROGRAM F043

File: problems/043/043.c
#include <stdio.h>
#include <stdlib.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    char main_num[11] = "0123456789";

    long int sum = 0;
    long int pandigital_num_list[100] = { 0 };

    int count = 0;

    char temp_char;
    for (int i = 0; i < 10; i++)
    {
        temp_char   = main_num[9];
        main_num[9] = main_num[8];
        main_num[8] = main_num[7];
        main_num[7] = main_num[6];
        main_num[6] = main_num[5];
        main_num[5] = main_num[4];
        main_num[4] = main_num[3];
        main_num[3] = main_num[2];
        main_num[2] = main_num[1];
        main_num[1] = main_num[0];
        main_num[0] = temp_char;

        for (int j = 0; j < 9; j++)
        {
            temp_char   = main_num[9];
            main_num[9] = main_num[8];
            main_num[8] = main_num[7];
            main_num[7] = main_num[6];
            main_num[6] = main_num[5];
            main_num[5] = main_num[4];
            main_num[4] = main_num[3];
            main_num[3] = main_num[2];
            main_num[2] = main_num[1];
            main_num[1] = temp_char;

            for (int k = 0; k < 8; k++)
            {
                temp_char   = main_num[9];
                main_num[9] = main_num[8];
                main_num[8] = main_num[7];
                main_num[7] = main_num[6];
                main_num[6] = main_num[5];
                main_num[5] = main_num[4];
                main_num[4] = main_num[3];
                main_num[3] = main_num[2];
                main_num[2] = temp_char;

                for (int l = 0; l < 7; l++)
                {
                    temp_char   = main_num[9];
                    main_num[9] = main_num[8];
                    main_num[8] = main_num[7];
                    main_num[7] = main_num[6];
                    main_num[6] = main_num[5];
                    main_num[5] = main_num[4];
                    main_num[4] = main_num[3];
                    main_num[3] = temp_char;

                    for (int m = 0; m < 6; m++)
                    {
                        temp_char   = main_num[9];
                        main_num[9] = main_num[8];
                        main_num[8] = main_num[7];
                        main_num[7] = main_num[6];
                        main_num[6] = main_num[5];
                        main_num[5] = main_num[4];
                        main_num[4] = temp_char;

                        for (int n = 0; n < 5; n++)
                        {
                            temp_char   = main_num[9];
                            main_num[9] = main_num[8];
                            main_num[8] = main_num[7];
                            main_num[7] = main_num[6];
                            main_num[6] = main_num[5];
                            main_num[5] = temp_char;

                            for (int o = 0; o < 4; o++)
                            {
                                temp_char   = main_num[9];
                                main_num[9] = main_num[8];
                                main_num[8] = main_num[7];
                                main_num[7] = main_num[6];
                                main_num[6] = temp_char;

                                for (int p = 0; p < 3; p++)
                                {
                                    temp_char   = main_num[9];
                                    main_num[9] = main_num[8];
                                    main_num[8] = main_num[7];
                                    main_num[7] = temp_char;

                                    for (int q = 0; q < 2; q++)
                                    {
                                        temp_char   = main_num[9];
                                        main_num[9] = main_num[8];
                                        main_num[8] = temp_char;

                                        char d3[2] = { main_num[2] };
                                        char d4[2] = { main_num[3] };
                                        char d5[2] = { main_num[4] };

                                        if ((main_num[5] != '5') && (main_num[5] != '0'))
                                            continue;

                                        if ((atoi(d4) % 2) != 0)
                                            continue;

                                        if (((atoi(d3) + atoi(d4) + atoi(d5)) % 3) != 0)
                                            continue;

                                        char d567[4]  = { main_num[4], main_num[5], main_num[6] };
                                        if ((atoi(d567) % 7) != 0)
                                            continue;

                                        char d678[4]  = { main_num[5], main_num[6], main_num[7] };
                                        if ((atoi(d678) % 11) != 0)
                                            continue;

                                        char d789[4]  = { main_num[6], main_num[7], main_num[8] };
                                        if ((atoi(d789) % 13) != 0)
                                            continue;

                                        char d8910[4] = { main_num[7], main_num[8], main_num[9] };
                                        if ((atoi(d8910) % 17) != 0)
                                            continue;

                                        int duplicate_found = 0;
                                        for (int r = 0; r < count; r++)
                                        {
                                            if (pandigital_num_list[r] == atol(main_num))
                                            {
                                                duplicate_found = 1;
                                                break;
                                            }
                                        }

                                        if (!duplicate_found)
                                        {
                                            pandigital_num_list[count++] = atol(main_num);
                                            sum += atol(main_num);
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    printf("%ld\n", sum);

    return 0;
}
